<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from kr2_robot_a810_tio.xacro       | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot features="TIO" mechanical-revision="5" model="a810" model-revision="051020-0" name="kr2_robot_a810">
  <!-- materials -->
  <material name="gray">
    <color rgba="0.5 0.5 0.5 1.0"/>
  </material>
  <material name="dark_gray">
    <color rgba="0.3 0.3 0.3 1.0"/>
  </material>
  <material name="light_gray">
    <color rgba="0.7 0.7 0.7 1.0"/>
  </material>
  <material name="black">
    <color rgba="0.1 0.1 0.1 1.0"/>
  </material>
  <material name="orange">
    <color rgba="1.0 0.4235294117647059 0.0392156862745098 1.0"/>
  </material>
  <!-- links -->
  <link name="base">
    <!--
    <xacro:visual_and_collision rpy="0 0 0"
                                xyz="0 0 ${base_outer_height / 2}"
                                material="dark_gray">
      <cylinder length="${base_outer_height}" radius="${base_outer_radius}"/>
    </xacro:visual_and_collision>

    <xacro:visual_and_collision rpy="0 0 0"
                                xyz="0 0 ${base_outer_height + base_inner_height/2}"
                                material="dark_gray">
      <cylinder length="${base_inner_height}" radius="${base_inner_radius}"/>
    </xacro:visual_and_collision>
    -->
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.6"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://kr_rviz_demo/meshes/a810/a810_Base.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
  </link>
  <link name="link1">
    <!--
    <xacro:visual_and_collision rpy="0 0 0"
                                xyz="0 0 ${link1_length/2 - 25 * scale}"
                                material="gray">
      <cylinder length="${link1_length - 50 * scale}" radius="${link1_radius}"/>
    </xacro:visual_and_collision>

    <xacro:visual_and_collision rpy="${pi/2} 0 0"
                                xyz="${50 * scale} ${20 * scale} ${link1_length}"
                                material="gray">
      <cylinder length="${link1_radius * 1.7}" radius="${link1_radius}"/>
    </xacro:visual_and_collision>
    
    <xacro:visual_and_collision rpy="0 ${pi/4} 0"
                                xyz="${20 * scale} ${+8 * scale} ${200 * scale}"
                                material="light_gray">
      <cylinder length="${90 * scale}" radius="${50 * scale}"/>
    </xacro:visual_and_collision>
    -->
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.13684999999999997"/>
      <mass value="4.58"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.169"/>
      <geometry>
        <mesh filename="package://kr_rviz_demo/meshes/a810/a810_Link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
  </link>
  <link name="linkJ2">
    <visual>
      <origin rpy="1.570796326795 0 0" xyz="0 -0.01 0"/>
      <geometry>
        <cylinder length="0.15" radius="0.015"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <origin rpy="1.570796326795 0 0" xyz="0 -0.01 0"/>
      <geometry>
        <cylinder length="0.15" radius="0.015"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="6.1"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
  </link>
  <link name="link2">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.17500000000000002"/>
      <mass value="0.43"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
    <visual>
      <origin rpy="-1.570796326795 0 -1.570796326795" xyz="0.049 0 0"/>
      <geometry>
        <mesh filename="package://kr_rviz_demo/meshes/a810/a810_Link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
  </link>
  <link name="link3">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="2.9"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
    <visual>
      <origin rpy="1.570796326795 0 -1.570796326795" xyz="-0.05 -0.0 0.315"/>
      <geometry>
        <mesh filename="package://kr_rviz_demo/meshes/a810/a810_Link4.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
  </link>
  <link name="linkJ4">
    <visual>
      <origin rpy="1.570796326795 0 0" xyz="0 0.013000000000000001 0"/>
      <geometry>
        <cylinder length="0.129" radius="0.015"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <origin rpy="1.570796326795 0 0" xyz="0 0.013000000000000001 0"/>
      <geometry>
        <cylinder length="0.129" radius="0.015"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0.006500000000000001 0"/>
      <mass value="3.8"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
  </link>
  <link name="link4">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.168"/>
      <mass value="0.33"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
    <visual>
      <origin rpy="-1.570796326795 0 1.570796326795" xyz="0.04 0.0005 0.0"/>
      <geometry>
        <mesh filename="package://kr_rviz_demo/meshes/a810/a810_Link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
  </link>
  <link name="link5">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1.7"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
    <visual>
      <origin rpy="0 0 -1.570796326795" xyz="0.0 0.002 0.1035"/>
      <geometry>
        <mesh filename="package://kr_rviz_demo/meshes/a810/a810_Link6.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
  </link>
  <link name="linkJ6">
    <!--
    <xacro:visual_and_collision rpy="${pi/2} 0 0" 
                                xyz="0 ${10 * scale} 0"
                                material="orange">
      <cylinder length="${cylinderJ6_height * 1.9}" radius="${cylinderJ6_radius}"/>
    </xacro:visual_and_collision>
    -->
    <inertial>
      <origin rpy="0 0 0" xyz="0 0.004 0"/>
      <mass value="1.8"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
  </link>
  <link name="link6">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.004"/>
      <mass value="1.8"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://kr_rviz_demo/meshes/a810/a810_Link7.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
  </link>
  <link name="end_effector">
    <visual>
      <origin rpy="0 0 0" xyz="0.0 -0.0 0.0"/>
      <geometry>
        <mesh filename="package://kr_rviz_demo/meshes/a810/a810_ToolIO.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.8 0.8 0.8 1"/>
      </material>
    </visual>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 -0.0197"/>
      <mass value="0.56"/>
      <!--      <inertia ixx="${mass/12*(3*radius*radius+3*length*length)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12*(3*radius*radius+3*length*length)}"
               iyz="0.0"
               izz="${mass*radius*radius/2}"/>
       -->
      <inertia ixx="0.003" ixy="0.0" ixz="0.0" iyy="0.003" iyz="0.0" izz="0.003"/>
    </inertial>
  </link>
  <!-- joints -->
  <joint name="joint0" type="fixed">
    <parent link="base"/>
    <child link="link1"/>
    <origin rpy="0 0 0" xyz="0 0 0.035"/>
    <axis xyz="0 0 1"/>
    <dynamics damping="0.1"/>
    <limit effort="204" lower="-6.28318530718" upper="6.28318530718" velocity="3.9269908169875"/>
    <!-- <limit effort="204" velocity="${225/180*pi}" lower="${-122.6 * pi_per_deg}" upper="${127.6 * pi_per_deg}"/> -->
  </joint>
  <joint name="joint1" type="fixed">
    <parent link="link1"/>
    <child link="linkJ2"/>
    <origin rpy="0 0 0" xyz="0.05 0.0 0.27369999999999994"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.1"/>
    <limit effort="204" lower="-1.221730476396111" upper="3.14159265359" velocity="3.9269908169875"/>
    <!-- <limit effort="204" velocity="${225/180*pi}" lower="${-31.9 * pi_per_deg}" upper="${168.5 * pi_per_deg}"/> -->
  </joint>
  <joint name="jointPL2" type="fixed">
    <parent link="linkJ2"/>
    <child link="link2"/>
    <origin rpy="0 0 0" xyz="-0.05 -0.0857 0"/>
  </joint>
  <joint name="joint2" type="fixed">
    <parent link="link2"/>
    <child link="link3"/>
    <origin rpy="0 0 0" xyz="0 0 0.274"/>
    <axis xyz="0 0 1"/>
    <dynamics damping="0.1"/>
    <limit effort="85" lower="-6.28318530718" upper="6.28318530718" velocity="3.9269908169875"/>
    <!--  <limit effort="85" velocity="${225/180*pi}" lower="${-177.6 * pi_per_deg}" upper="${177.6 * pi_per_deg}"/> -->
  </joint>
  <joint name="joint3" type="fixed">
    <parent link="link3"/>
    <child link="linkJ4"/>
    <origin rpy="0 0 0" xyz="0.04 0.0 0.1429"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.1"/>
    <limit effort="85" lower="-1.221730476396111" upper="3.14159265359" velocity="3.9269908169875"/>
    <!-- <limit effort="85" velocity="${225/180*pi}" lower="${-20.6 * pi_per_deg}" upper="${152.1 * pi_per_deg}"/> -->
  </joint>
  <joint name="jointPL4" type="fixed">
    <parent link="linkJ4"/>
    <child link="link4"/>
    <origin rpy="0 0 0" xyz="-0.04 0.08405 0.0"/>
  </joint>
  <joint name="joint4" type="fixed">
    <parent link="link4"/>
    <child link="link5"/>
    <origin rpy="0 0 0" xyz="0 0 0.271"/>
    <axis xyz="0 0 1"/>
    <dynamics damping="0.1"/>
    <limit effort="51" lower="-6.28318530718" upper="6.28318530718" velocity="3.9269908169875"/>
    <!-- <limit effort="51" velocity="${225/180*pi}" lower="${-177.6 * pi_per_deg}" upper="${177.6 * pi_per_deg}"/> -->
  </joint>
  <joint name="joint5" type="fixed">
    <parent link="link5"/>
    <child link="linkJ6"/>
    <origin rpy="0 0 0" xyz="0 0 0.10350000000000001"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.1"/>
    <limit effort="51" lower="-6.28318530718" upper="6.28318530718" velocity="3.9269908169875"/>
    <!--    <limit effort="51" velocity="${225/180*pi}" lower="${-177.6 * pi_per_deg}" upper="${177.6 * pi_per_deg}"/> -->
  </joint>
  <joint name="jointPL6" type="fixed">
    <parent link="linkJ6"/>
    <child link="link6"/>
    <origin rpy="0 0 0" xyz="0 0.111 0"/>
  </joint>
  <joint name="joint6" type="fixed">
    <parent link="link6"/>
    <child link="end_effector"/>
    <origin rpy="0 0 0" xyz="0 0 0.108"/>
    <axis xyz="0 0 1"/>
    <dynamics damping="0.1"/>
    <limit effort="51" lower="-6.28318530718" upper="6.28318530718" velocity="3.9269908169875"/>
    <!-- <limit effort="51" velocity="${225/180*pi}" lower="${-177.6 * pi_per_deg}" upper="${177.6 * pi_per_deg}"/> -->
  </joint>
  <!-- gazebo -->
  <!-- Used for fixing robot to Gazebo 'base_link' -->
  <link name="world"/>
  <joint name="joint_world" type="fixed">
    <parent link="world"/>
    <child link="base"/>
  </joint>
  <gazebo>
    <plugin filename="libgazebo_ros_control.so" name="ros_control">
      <robotNamespace>/kr2_robot</robotNamespace>
    </plugin>
  </gazebo>
  <gazebo reference="base">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="link1">
    <material>Gazebo/Grey</material>
  </gazebo>
  <gazebo reference="link2">
    <material>Gazebo/White</material>
  </gazebo>
  <gazebo reference="link3">
    <material>Gazebo/Grey</material>
  </gazebo>
  <gazebo reference="link4">
    <material>Gazebo/White</material>
  </gazebo>
  <gazebo reference="link5">
    <material>Gazebo/Grey</material>
  </gazebo>
  <gazebo reference="link6">
    <material>Gazebo/White</material>
  </gazebo>
  <gazebo reference="end_effector">
    <material>Gazebo/Black</material>
  </gazebo>
  <!-- Transmission mappings -->
  <transmission name="tran0">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint0">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor0">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint1">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint2">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor2">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint3">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor3">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint4">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor4">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran5">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint5">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor5">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran6">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint6">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor6">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
</robot>
